File: leetcode/easy/112-path-sum/recursive.cpp
/**
 * @file recursive.cpp
 * @author Vortexx2
 * @brief Problem 112 - Path Sum
 * @date 27-07-2021
 * 
 * Runtime - 4 ms
 * Memory Usage - 21.3 MB
 */
#include <iostream>
#include <vector>
#include <deque>
#include <algorithm>
#include <list>
#include <unordered_map>
#include <map>
#include <set>
#include <stack>
#include <queue>
#include <string>

using namespace std;

struct TreeNode
{
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode() : val(0), left(nullptr), right(nullptr) {}
  TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
  TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution
{
public:
  bool hasPathSum(TreeNode *root, int targetSum)
  {
    if (!root)
      return false;

    if (!root->left && !root->right && targetSum == root->val)
      return true;

    int a = targetSum - root->val;
    return hasPathSum(root->left, a) || hasPathSum(root->right, a);
  }
};

File: leetcode/easy/112-path-sum/recursive.py
"""
  @author Vortexx2
  Problem 112 - Path Sum

  Runtime - 36 ms
  Memory Usage - 15.8 MB
"""
class TreeNode:
  def __init__(self, val=0, left=None, right=None):
    self.val = val
    self.left = left
    self.right = right


class Solution:
  def hasPathSum(self, root: TreeNode, targetSum: int) -> bool:

    if not root:
      return False

    if (not root.left and not root.right and targetSum == root.val):
      return True

    a = targetSum - root.val
    return self.hasPathSum(root.left, a) or self.hasPathSum(root.right, a)
